File: scripts/check-th-search-r1-001-browser.py
"""Bounded, operator-invoked live browser proof; never run by ordinary CI.

Requires an already authorized agent-browser session and its executable path.
Only public research questions are captured. No data writes or auth state dumps.
"""
import argparse, base64, datetime, json, pathlib, subprocess, urllib.parse

p = argparse.ArgumentParser()
p.add_argument('--browser', required=True)
p.add_argument('--origin', required=True)
p.add_argument('--label', required=True)
p.add_argument('--sha', required=True)
p.add_argument('--deployment', required=True)
args = p.parse_args()
folder = pathlib.Path('docs/qa/th-search-r1-001')
report = {'utc': datetime.datetime.now(datetime.timezone.utc).isoformat(), 'surface': 'persistent Chromium via agent-browser', 'origin': args.origin, 'sha': args.sha, 'deployment': args.deployment, 'checks': []}

def browser(*argv):
    run = subprocess.run([args.browser, '--session', 'th-r1', '--json', *argv], capture_output=True, text=True, encoding='utf-8', timeout=65)
    data = json.loads(run.stdout)
    if run.returncode or not data.get('success'): raise RuntimeError(str(data))
    return data.get('data', {})

def js(code):
    return browser('eval', '-b', base64.b64encode(code.encode()).decode()).get('result')

def settled():
    browser('wait', '--load', 'networkidle')
    return js("({path:location.pathname,text:document.querySelector('main').innerText,width:innerWidth,scroll:document.documentElement.scrollWidth,robots:document.querySelector('meta[name=robots]')?.content})")

def screenshot(name):
    browser('screenshot', str((folder / (args.label+'-'+name+'.png')).resolve()))

try:
    for width, question in [(1280, 'USDOT 3244649'), (390, 'usdot 3244 649 miama movers')]:
        browser('set', 'viewport', str(width), '900')
        browser('open', args.origin+'/')
        browser('wait', '--load', 'networkidle')
        browser('wait', '3000')
        browser('find', 'label', 'Mover research question, company, USDOT or MC number', 'fill', question)
        assert js('document.activeElement.value') == question
        browser('press', 'Enter')
        browser('wait', '--url', '**/ask?*')
        ui = settled()
        assert ui['path']=='/ask' and 'SHIFL INC' in ui['text']
        assert 'records USDOT 3244649.' in ui['text'] and 'records USDOT 3244.' not in ui['text']
        assert ui['scroll'] <= width
        assert 'noindex' in ui['robots']
        if width==390: assert 'miama movers' in ui['text'] and 'Not established' in ui['text']
        # Keyboard-operated native disclosure, with real browser focus and Enter.
        for name in ['Trace this result', 'Trace this query']:
            js("(()=>{const e=[...document.querySelectorAll('summary')].find(e=>e.textContent.trim()==="+json.dumps(name)+");e.focus();e.scrollIntoView({block:'center'});return document.activeElement===e})()")
            browser('press', 'Enter')
            assert js("[...document.querySelectorAll('summary')].find(e=>e.textContent.trim()==="+json.dumps(name)+").parentElement.open")
        ui = settled()
        assert 'USDOT 3244649' in ui['text'] and 'MC 1019808' in ui['text']
        assert ui['scroll'] <= width
        api = js("fetch('/api/ask?q='+encodeURIComponent("+json.dumps(question)+")).then(async r=>({status:r.status,data:await r.json()}))")
        assert api['status']==200 and len(api['data']['results'])==1
        row=api['data']['results'][0]
        assert row['usdot']=='3244649' and row['mc']=='1019808'
        assert row['name'] in ui['text'] and row['whyMatched'] in ui['text']
        assert api['data']['counts'][0]['value']==1
        js("[...document.querySelectorAll('h3')].find(e=>e.textContent==='SHIFL INC').scrollIntoView({block:'center'})")
        screenshot('identity-'+str(width))
        report['checks'].append({'name':'homepage-keyboard-native-api-trace', 'width':width,'question':question,'ui':ui,'api':api})
    browser('set','viewport','320','844')
    ui=settled(); assert ui['scroll']<=320
    screenshot('identity-320')
    report['checks'].append({'name':'320-overflow','width':320,'scroll':ui['scroll']})
    questions = ['USDOT 3244 649','MC 1019 808','USDOT 3244','USDOT lookup','USDOT 3244 MC 649','mover in new jersey','How many current household-goods carriers are headquartered in New Jersey?','NJ PM/PW/PC licensed movers','movers in Miami']
    for index, question in enumerate(questions):
        browser('set','viewport', '1280' if index%2 else '390','900')
        browser('open',args.origin+'/ask?q='+urllib.parse.quote(question))
        ui=settled(); assert ui['path']=='/ask' and ui['scroll']<=ui['width']
        api=js("fetch('/api/ask?q='+encodeURIComponent("+json.dumps(question)+")).then(async r=>({status:r.status,data:await r.json()}))")
        data=api['data']; assert api['status']==200
        assert data['terminalState']!='UNAVAILABLE'
        for row in data['results']: assert row['name'] in ui['text'] and row['whyMatched'] in ui['text']
        if index in [0,1]: assert len(data['results'])==1 and data['results'][0]['usdot']=='3244649'
        if index==2: assert all(r['usdot']=='3244' for r in data['results'])
        if index==3: assert data['terminalState']=='NEEDS_CLARIFICATION'
        if index==4: assert not data['results']
        if index==5: assert data['results'] and 'Recorded' in ui['text'] and data['query']['executor']=='directory'
        if index==6:
            assert data['query']['mode']=='count' and data['query']['role']=='carrier' and data['query']['authorityCurrent'] is True
            assert data['query']['jurisdiction']['state']=='NJ' and data['counts']
            assert str(data['counts'][0]['value']) in ui['text']
        if index in [7,8]: assert not data['results'] and data['coverageState'] in ['UNSUPPORTED','REQUEST_ONLY','NOT_ACQUIRED']
        if index in [3,5,6,7]: screenshot('case-'+str(index))
        report['checks'].append({'name':question,'ui':ui,'api':api})
    report['status']='PASS'
finally:
    (folder/(args.label+'-browser.json')).write_text(json.dumps(report,indent=2),encoding='utf-8')
print(json.dumps({'status':report.get('status','FAIL'),'checks':len(report['checks']),'artifact':str(folder/(args.label+'-browser.json'))}))

File: scripts/check-th-search-r1-001-browser.mjs
/** Operator-only bounded live QA. Attach to an authorized persistent browser.
 * Usage: PLAYWRIGHT_MODULE=<installed playwright-core> node this-file
 *   <cdp-local-url> <origin> <label> <sha> <deployment>
 * No writes, user state dumps, or live CI calls. */
import assert from 'node:assert/strict';
import {createRequire} from 'node:module';
import {writeFileSync} from 'node:fs';
import {resolve} from 'node:path';
const {chromium}=createRequire(import.meta.url)(process.env.PLAYWRIGHT_MODULE || 'playwright-core');
const [cdp,origin,label,sha,deployment]=process.argv.slice(2);
const browser=await chromium.connectOverCDP(cdp);
const page=browser.contexts()[0].pages().find(p=>p.url().startsWith(origin)) || await browser.contexts()[0].newPage();
const folder='docs/qa/th-search-r1-001';
const report={utc:new Date().toISOString(),surface:'Persistent Chrome, Playwright attached via local CDP',origin,sha,deployment,checks:[]};
async function hydrated(){await page.waitForFunction(()=>{const e=document.querySelector('input[name=q]');return e&&Object.keys(e).some(k=>k.startsWith('__reactProps'))});}
async function ui(){await page.waitForLoadState('networkidle');await hydrated();return page.evaluate(()=>({path:location.pathname,text:document.querySelector('main').innerText,width:innerWidth,scroll:document.documentElement.scrollWidth,robots:document.querySelector('meta[name=robots]')?.content}));}
async function api(q){return page.evaluate(async q=>{const r=await fetch('/api/ask?'+new URLSearchParams({q}));return{status:r.status,data:await r.json()}},q);}
async function snap(name){await page.screenshot({path:resolve(folder,`${label}-${name}.png`)});}
try {
  for(const [width,q] of [[1280,'USDOT 3244649'],[390,'usdot 3244 649 miama movers']]){
    await page.setViewportSize({width,height:900});await page.goto(origin+'/');await page.waitForLoadState('networkidle');await hydrated();
    const input=page.getByLabel('Mover research question, company, USDOT or MC number');await input.evaluate(e=>e.scrollIntoView({behavior:'instant',block:'center'}));await input.click();await input.fill(q);await input.press('Enter');await page.waitForURL('**/ask?*');
    let view=await ui();assert.equal(view.path,'/ask');assert.ok(view.text.includes('SHIFL INC'));assert.ok(view.text.includes('records USDOT 3244649.'));assert.ok(!view.text.includes('records USDOT 3244.'));assert.ok(view.scroll<=width);assert.match(view.robots,/noindex/);
    if(width===390){assert.match(view.text,/miama movers/);assert.match(view.text,/Not established/);}
    for(const name of ['Trace this result','Trace this query']){
      const summary=page.locator('summary').filter({hasText:name});await summary.evaluate(e=>e.scrollIntoView({behavior:'instant',block:'center'}));await page.waitForTimeout(300);const original=await summary.elementHandle();await summary.click();assert.ok(await original.evaluate(e=>e.isConnected));assert.ok(await summary.evaluate(e=>e.parentElement.open));
      await summary.press('Enter');assert.equal(await summary.evaluate(e=>e.parentElement.open),false);
      await summary.press('Enter');assert.ok(await summary.evaluate(e=>e.parentElement.open));
    }
    view=await ui();assert.match(view.text,/MC 1019808/);assert.ok(view.scroll<=width);
    const response=await api(q);assert.equal(response.status,200);assert.equal(response.data.results.length,1);const row=response.data.results[0];assert.equal(row.usdot,'3244649');assert.equal(row.mc,'1019808');assert.ok(view.text.includes(row.whyMatched));assert.equal(response.data.counts[0].value,1);
    await page.getByRole('heading',{name:'SHIFL INC',exact:true}).scrollIntoViewIfNeeded();await snap('identity-'+width);
    report.checks.push({name:'homepage-keyboard-native-api-trace',width,q,ui:view,api:response});
  }
  await page.setViewportSize({width:320,height:844});const narrow=await ui();assert.ok(narrow.scroll<=320);await snap('identity-320');report.checks.push({name:'320-overflow',width:320,scroll:narrow.scroll});
  const questions=['USDOT 3244 649','MC 1019 808','USDOT 3244','USDOT lookup','USDOT 3244 MC 649','mover in new jersey','How many current household-goods carriers are headquartered in New Jersey?','NJ PM/PW/PC licensed movers','movers in Miami','current interstate movers headquartered in Florida'];
  for(const [i,q] of questions.entries()){
    await page.setViewportSize({width:i%2?1280:390,height:900});await page.goto(origin+'/ask?'+new URLSearchParams({q}));const view=await ui();assert.equal(view.path,'/ask');assert.ok(view.scroll<=view.width);
    const response=await api(q),data=response.data;assert.equal(response.status,200);assert.notEqual(data.terminalState,'UNAVAILABLE');
    for(const row of data.results){assert.ok(view.text.includes(row.name));assert.ok(view.text.includes(row.whyMatched));}
    if(i<2){assert.equal(data.results.length,1);assert.equal(data.results[0].usdot,'3244649');}
    if(i===2)assert.ok(data.results.every(r=>r.usdot==='3244'));
    if(i===3)assert.equal(data.terminalState,'NEEDS_CLARIFICATION');
    if(i===4)assert.equal(data.results.length,0);
    if(i===5){assert.ok(data.results.length);assert.match(view.text,/recorded/i);assert.equal(data.query.executor,'directory');}
    if(i===6){assert.equal(data.query.mode,'count');assert.equal(data.query.role,'carrier');assert.equal(data.query.authorityCurrent,true);assert.equal(data.query.jurisdiction.state,'NJ');assert.ok(data.counts.length);assert.ok(view.text.includes(String(data.counts[0].value)));}
    if([7,8].includes(i)){assert.equal(data.results.length,0);assert.ok(['UNSUPPORTED','REQUEST_ONLY','NOT_ACQUIRED'].includes(data.coverageState));}
    if(i===9)assert.ok(data.results.length);
    if([3,5,6,7].includes(i))await snap('case-'+i);
    report.checks.push({name:q,ui:view,api:response});
  }
  report.status='PASS';
} finally {
  writeFileSync(`${folder}/${label}-browser.json`,JSON.stringify(report,null,2)+'\n');await browser.close();
}
console.log(JSON.stringify({status:report.status,checks:report.checks.length,artifact:`${folder}/${label}-browser.json`}));
